### PR TITLE
Implement basic URL shortening service with API endpoints

### PR DESCRIPTION
**Pull request by Premodial**
# Basic URL Shortening Service Implementation

This Pull Request introduces a basic URL shortening service using Node.js. The service includes API endpoints for creating shortened URLs and redirecting to the original URLs.

## Key Features

- **Express.js Server**: The web server is set up using Express.js for efficient handling of API requests.
- **Database Integration**: A database schema is created to store mappings of original to shortened URLs.
- **Unique Shortening Logic**: Implemented logic ensures that each shortened URL is unique.
- **API Endpoints**:
  - `POST /shorten`: Endpoint to submit a URL to be shortened.
  - `GET /:shortened`: Endpoint that redirects users from the shortened URL to the original URL.
- **Error Handling**: Robust error handling for invalid inputs and server errors.
- **URL Validation**: Implemented validation to ensure URLs are in a correct format before shortening.

## How to Set Up and Run the Project

1. Clone the repository to your local machine.
2. Run `npm install` to install dependencies.
3. Start the server with `npm run start`.
4. The server will run on `http://localhost:3000`.

## How to Use the API

- To shorten a URL, send a `POST` request to `/api/shorten` with a JSON body containing the URL.
- Access the shortened URL by visiting `http://localhost:3000/api/:shortened` in a web browser, which will redirect you to the original URL.

## Running Tests

Execute `npm run test` to run unit tests and ensure that all functionalities are working as expected.
The server need to be running for the test to run.

### CODE REVIEW

File: src/controllers/urlController.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+ *
+ * @param {Request} req - Express request object, containing the original URL in the body.
+ * @param {Response} res - Express response object for sending back the shortened URL or errors.
+ * @returns {Promise<Response>} A promise resolving with the operation result.
+ */
+export const shortenUrl = async (req: Request, res: Response): Promise<Response> => {
+  try {
+    const { originalUrl } = req.body;
+    if (!validateUrl(originalUrl)) {
+      return res.status(400).json({ error: 'Invalid URL format.' });
```
Comment (Premodial) [inline src/controllers/urlController.ts:29]: Could we add a more informative error message here, specifying why the URL format is invalid?

File: src/routes/index.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import { Router } from 'express';
+import { shortenUrl, redirectUrl } from '../controllers/urlController';
```
Comment (Premodial) [inline src/routes/index.ts:2]: Great use of the Express Router for modularity! 💫💫💫

File: src/app.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+const PORT = process.env.PORT || 3000;
+
+/**
+ * Middleware registration.
+ * - bodyParser.json() is used to parse incoming request bodies in a middleware before your handlers, available under the req.body property.
+ * - bodyParser.urlencoded({ extended: true }) is used to parse URL-encoded bodies and allows rich objects and arrays to be encoded into the URL-encoded format, allowing for a JSON-like experience with URL-encoded.
+ */
+app.use(bodyParser.json());
+app.use(bodyParser.urlencoded({ extended: true }));
```
Comment (Premodial) [inline src/app.ts:21]: Consider removing the usage of `bodyParser.json()` and `bodyParser.urlencoded()` since Express has built-in support for these middlewares via `express.json()` and `express.urlencoded()`. This would not only simplify the code but also reduce redundancy. Great job leveraging environment variables for the port configuration, enhancing the app's flexibility across different environments!

File: src/controllers/urlController.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    const urlCode = generateUniqueId(10); // You can adjust the length as needed
+    const existingUrls = await urlsCollection.where('originalUrl', '==', originalUrl).get();
+
```
Comment (Premodial) [inline src/controllers/urlController.ts:34]: While checking for existing URLs to prevent duplication is efficient, consider if the requirements specify whether the same original URL should always result in the same shortened URL. If unique short URLs are desired for each submission of the same original URL, this logic may need to be adjusted. Excellent job structuring the async flow with try-catch for robust error handling!✨

File: src/firebaseAdmin.ts [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+const serviceAccount = require('./config/service-account.json');
+
+// Initialize Firebase Admin SDK.
+admin.initializeApp({
+  credential: admin.credential.cert(serviceAccount as admin.ServiceAccount),
+  // Specify the storage bucket URL here. This is used for Firebase Storage operations.
+  storageBucket: 'gs://weather-app-f801d.appspot.com',
+});
+
```
Comment (Premodial) [inline src/firebaseAdmin.ts:11]: Securing the Firebase configuration by abstracting it into a separate file is a commendable practice✨✨✨. As a suggestion, ensure that the `service-account.json` file and details like `storageBucket` are securely managed and not hard-coded, possibly using environment variables. This approach enhances security and configurability.

File: src/services/urlService.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+ * @returns {boolean} True if the URL is valid, false otherwise. This boolean response enables
+ * straightforward checking within controllers or other parts of the application that require URL validation.
+ */
+export const validateUrl = (url: string): boolean => {
+  return validator.isURL(url, {
+    protocols: ['http','https'],
+    require_protocol: true
+  });
+};
+
```
Comment (Premodial) [inline src/services/urlService.ts:23]: Utilizing the `validator` library for URL validation is a solid choice for ensuring that input URLs meet a basic format requirement💫💫. To further enhance security and functionality, consider extending the validation logic to exclude potentially malicious URLs or to support additional protocols if needed. Your attention to detail in input validation is a key strength of this service.

File: test/urlService.test.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
+// tests/urlShorteningService.test.ts
+import request from 'supertest';
+const BASE_URL = 'http://localhost:3000';
+
+describe('URL Shortening Service Integration Tests', () => {
+  /**
+   * Tests the scenario where a new URL is shortened.
+   * Expects a 200 OK response with a defined shortUrl in the response body.
+   */
+  it('successfully shortens a new URL', async () => {
```
Comment (Premodial) [inline test/urlService.test.ts:10]: Writing integration tests for the URL shortening and redirection functionality is crucial for ensuring the reliability of your service. For an even more robust testing strategy, consider mocking the database calls to simulate various scenarios and responses. This approach would allow you to test edge cases and error handling more comprehensively without depending on a live database connection.

File: test/urlService.test.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+  /**
+   * Tests that accessing a non-existent short URL returns a 404 Not Found status.
+   */
+  it('returns 404 for an invalid shortId', async () => {
+    const response = await request(BASE_URL).get('/nonexistentShortId');
+
+    expect(response.statusCode).toBe(404); // Ensures the server correctly responds to invalid short URLs.
+  });
+});
```
Comment (Premodial) [inline test/urlService.test.ts:58]: Fantastic work on this URL shortening service! Your followed the best practices in project setup, database integration, and API design sets a solid foundation for a reliable and scalable application. Your thoughtful implementation of error handling and validation significantly enhances the user experience and security of the service. Moving forward, consider the suggestions regarding redundancy elimination, unique URL generation logic, secure configuration management, validation extension, and test enhancement to further refine and improve the service. Keep up the great work! 💫💫💫
